### PR TITLE
feat: Bulk Download List Endpoints

### PR DESCRIPTION
**Pull request by rainandbare**
# Pull Request

## JIRA Ticket

[CZID-8831](https://czi-tech.atlassian.net/browse/CZID-8831)
[CZID-8832](https://czi-tech.atlassian.net/browse/CZID-8832)

## Description

Add Federation support for:
- GET /bulk_downloads/{bulkDownloadId}.json
- GET /bulk_downloads.json
- GET /bulk_downloads/{bulkDownloadId}/presigned_output_url

## Tests

Test GraphQL with this call:
```
query MyQuery {
  bulkDownloads(input:{
    limit:2,
    searchBy: "Suzette McCanny"
  }){
    id
    status
    startedAt
    ownerUserId
    analysisCount
    entityInputFileType
    sampleNames
    rawInputsJson {
      description
      downloadType
      downloadDisplayName
      fileFormat
    }
    file {
      downloadLink {
        url
      }
      size
    }
    toDelete {
      user_name
      log_url
      totalSamples
      progress
    }
  }
```

[CZID-8831]: https://czi-tech.atlassian.net/browse/CZID-8831?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CZID-8832]: https://czi-tech.atlassian.net/browse/CZID-8832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .meshrc.yaml [hunk 1]
```diff
@@ -39,13 +39,15 @@ sources:
             responseSchema: ./json-schemas/snapshotBackground.json
             responseTypeName: Background
           - type: Query
-            field: BulkDownloadDetails
-            path: /bulk_downloads/{args.bulkDownloadId}
```
Comment (bchu1) [inline .meshrc.yaml:43]: does `path` actually do anything, since we use custom resolvers? I've been curious about this...

File: resolvers.ts [hunk 2]
```diff
@@ -48,6 +48,98 @@ export const resolvers: Resolvers = {
...
+      const statusDictionary = {
+        "success":"SUCCEEDED",
+        "error": "FAILED",
+        "waiting": "PENDING",
+        "running": "INPROGRESS",
+        //fyi: in NextGen there is also a status of STARTED
+      }
+      const urlParams = formatUrlParams({
+        searchBy: args?.input?.searchBy,
+        n:args?.input?.limit
```
Comment (bchu1) [inline resolvers.ts:61]: nit - space after `n`

File: resolvers.ts [hunk 3]
```diff
@@ -48,6 +48,98 @@ export const resolvers: Resolvers = {
...
+          user_name, 
+          log_url,
+          analysis_type,
+          progress // --> to be discussed on Feb 16th, 2024
+          } = bulkDownload;
+
+          // In Next Gen we will have an array with all of the entity input 
+          // filtered through the nodes entity query to get the relevant info
+          // If there are 22 Consensus Genome Files coming from 20 Samples, there will be 42 items in the array.
+          // We will get `sampleNames` by checking __typename to see if the entity is a sample,
```
Comment (bchu1) [inline resolvers.ts:106]: is there a NextGen API that like returns a polymorphic type?

File: utils/httpUtils.ts [hunk 4]
```diff
@@ -7,12 +7,18 @@ import fetch from "node-fetch";
  *  { items: [1, 2] } would become "items[]=1&items[]=2".
  */
 export const formatUrlParams = (params: { [s: string]: unknown }) => {
+  const replaceSpaces = (value: string) => {
```
Comment (bchu1) [inline utils/httpUtils.ts:10]: could we separate this out into a module scoped function?

File: utils/httpUtils.ts [hunk 5]
```diff
@@ -24,14 +30,15 @@ export const get = async (url: string, args: any, context: any) => {
   try {
     const baseURL = process.env.API_URL;
     const urlPrefix = args.snapshotLinkId ? `/pub/${args.snapshotLinkId}` : "";
+    console.log(baseURL + urlPrefix + url)
```
Comment (bchu1) [inline utils/httpUtils.ts:33]: console log

**Review by bchu1 — APPROVED** [review]

File: .meshrc.yaml (hunk 1 above)
Comment (rainandbare) [reply .meshrc.yaml:43]: lol - yeah - if you don't have custom resolvers than thats where it sends the query and I think there is a difference in how the arguments come into the query (if they are mentioned in the path then they are just basic but if they are in the resolver they come in through the input argument) - I haven't tried just leaving it out for custom resolvers thou.

File: resolvers.ts (hunk 3 above)
Comment (rainandbare) [reply resolvers.ts:106]: I was thinking something like this:
`nodes(ids: "") {
    _id
    ... on ConsensusGenome {
      id
    }
    ... on GenomicRange {
      id
    }
    ... on Sample {
      id
      name
    }
  }`

Comment (bchu1) [reply resolvers.ts:106]: I see

**Pull request merged**